Refactor solver schemes

gassmoeller opened the pull request:
This is an early stab at #972 and #1279, and is blocked until #1707 is merged. It is not ready for review, I open it mostly to take a look at the test results.

gassmoeller commented:
I have updated this PR to include the changes of the Newton solver scheme, so this should be ready for a review. In essence this PR does the following things:

- It moves the nonlinear solver schemes into their own functions to make the `solve_timestep` function easier to read.
- It unifies the assembly and solution of the single equations by combining them in the `assemble_and_solve_...` functions. This makes sure every solver scheme interprets the residuals the same way (and removes a source of error when writing a new solver scheme).
- It makes some functionality that was so far only implemented for some of the solver schemes available to all (e.g. free surface, operator splitting, active particles)

All of that should be achieved without changing tests, this is simply a (complicated) reshuffling of code that should hopefully make the flow of the code easier to read.
There are still some totally unreasonable decisions in some of the solver schemes, in particular concerning the conditions on when to reassemble the Stokes matrix and preconditioner. I will open a follow up PR to address those, because I can not unify them without changing tests.

MFraters commented on source/simulator/solver_schemes.cc:
```diff
@@ -0,0 +1,744 @@
+        const double relative_stokes_residual =
+          assemble_and_solve_stokes(nonlinear_iteration == 0, &initial_stokes_residual);
+
+        pcout << "      Relative Stokes residual after nonlinear iteration " << nonlinear_iteration+1
```
Not sure if it should be changed in this pull request or in a follow up, but since it is about unification, I will bring it up here anyway.  

I think we should also unify these outputs, so that it is easier to grep them. Some for example start with 'Total relative ... 'and others with 'Relative ... '. This could then for example become something like 'Nonlinear residual (Stokes) for iteration ... ' or 'Nonlinear residual (Total) for iteration ...' etc. And then always for every timestep end with Final residual for timestep ...

MFraters commented on include/aspect/simulator.h:
```diff
@@ -431,6 +431,90 @@ namespace aspect
+       *
+       * 'Newton Stokes' solves the temperature and composition equations
+       * once and then iterates over the Stokes equation until convergence
+       * is reached. It begins this iteration with Picard iterations and
```
Maybe add that it is a defect correction form of the Picard iteration.

gassmoeller commented:
Thanks for the comments! I addressed the first two. 
Could you make a separate issue out of the third comment? I know that @jdannberg looked at this in #1433, and I only remember it is complicated to make the output correct, precise and uniform for all schemes. But there may be better output schemes than we use right now, so feel free to suggest improvements.
About the fourth comment: I would prefer to keep it as is, I feel that nested max and min statements are kind of hard to read. I realize though that is a matter of taste, I hope you are ok with me keeping the current form.

gassmoeller commented:
Since @MFraters mentioned he would like somebody else to look at this, maybe @jdannberg or @tjhei could also take a look at this PR? You know probably most about our current solver schemes and in which direction you would like to see this going in the future.

MFraters commented:
Yes, I will make an issue for that. 
About your comment to my fourth comment: sure, I agree it is a bit easier to read this way, so I am fine with it staying this way.

bangerth commented on source/simulator/solver_schemes.cc:
```diff
@@ -0,0 +1,744 @@
+          }
+
+        max = std::max(relative_stokes_residual, max);
+        max = std::max(relative_temperature_residual, max);
```
Can't wait till we can finally use C++11 in ASPECT -- because then we can write `std::max(a,b,c)`!

bangerth commented on include/aspect/simulator.h:
```diff
@@ -431,6 +431,91 @@ namespace aspect
+       * 'IMPES' is the classical IMplicit
+       * Pressure Explicit Saturation scheme in which ones solves
+       * the temperatures and Stokes equations exactly
+       * once per time step, one after the other.
```
What a shame that we didn't give this a better name -- because what we actually implement these days is an IMPIS scheme since we also treat the temperature equations implicitly :-(

bangerth commented on include/aspect/simulator.h:
```diff
@@ -459,6 +544,45 @@ namespace aspect
+       * <code>source/simulator/solver_schemes.cc</code>.
+       */
+      double assemble_and_solve_temperature (const bool compute_initial_residual = false,
+                                             double *initial_residual = NULL);
```
The two arguments are duplicative: I think you can get rid of the first and document that the initial residual is computed if the pointer is non-`NULL`. Same for the other two functions of similar type.

gassmoeller replied:
I do not quite agree. Of course I could replace 
`if (compute_initial_residual)` 
by 
`if(initial_residual!=NULL)`, 
but then I would need to call the function differently, namely instead of 
```
assemble_and_solve_temperature(nonlinear_iteration==0,initial_residual)
``` 
I would need to call 
```
assemble_and_solve_temperature((nonlinear_iteration==0) ?initial_residual : NULL))
```
, otherwise I would recompute the initial residual in every iteration. But then I would not have the initial residual available inside of the function in later iterations, so I could not return the relative residual, but only the absolute residual. This means I would need to move the conversion to relative residual into the individual solver schemes, which opens up the possibility for the output of solver schemes to diverge. I guess in the end both is possible, but I would prefer the current way. Do you insist on the change?

bangerth replied:
No. I hadn't seen how you called the function. I think `(nonlinear_iteration==0) ?initial_residual : NULL)` would not be unreadable, but either is fine by me.

I did think about using `boost::optional` but decided that maybe that is going too far ;-)

bangerth commented on source/simulator/core.cc:
```diff
@@ -1838,696 +1747,53 @@ namespace aspect
         current_linearization_point = distr_solution;
       }
 
+    // The free surface scheme is currently not build to work inside a nonlinear solver.
```
build -> built

bangerth commented on source/simulator/solver_schemes.cc:
```diff
@@ -0,0 +1,744 @@
+  namespace
+  {
+    /**
+     * Converts a function with a certain number of components into a Function@dim@
```
-> Function<dim> or Function@<dim@>

bangerth commented on source/simulator/solver_schemes.cc:
```diff
@@ -0,0 +1,744 @@
+    return 0.0;
+  }
+
+  template <int dim>
```
add a couple more empty lines here

bangerth commented on source/simulator/solver_schemes.cc:
```diff
@@ -0,0 +1,744 @@
+    std::vector<double> current_residual(introspection.n_compositional_fields,0.0);
+
+    if (compute_initial_residual)
+      Assert(initial_residual->size() == introspection.n_compositional_fields, ExcNotImplemented());
```
in the current scheme, you probably also want to test that `initial_residual` is non-`NULL`. But see also my other comment further up.

bangerth replied:
Also, shouldn't it be an `ExcInternalError`? I mean, this shouldn't happen, right?

bangerth commented on source/simulator/solver_schemes.cc:
```diff
@@ -0,0 +1,744 @@
+
+
+  template <int dim>
+  double Simulator<dim>::assemble_and_solve_stokes (const bool compute_initial_residual,
```
I think you use uppercase `Stokes` for some of these functions, but this one is lowercase `stokes`. Maybe make it uniform?

bangerth replied:
I think we tended to use lower-case `stokes` everywhere. Maybe that suggests also using `newton_stokes`?

bangerth commented on source/simulator/solver_schemes.cc:
```diff
@@ -0,0 +1,744 @@
+    assemble_and_solve_stokes();
+
+    if (parameters.run_postprocessors_on_nonlinear_iterations)
+      postprocess ();
```
Is this function (`solve_IMPES`) called in a nonlinear loop? If not, shouldn't we always call this?

(I guess my question must be dumb because surely this would break a million output files otherwise...)

gassmoeller replied:
This function is not called inside a nonlinear loop, but the execution of postprocessor in the `run` function is disabled if we run them on every nonlinear iteration. The responsibility of calling them is transferred to the solver scheme in that case. Thus, even for solver schemes that do not have a nonlinear loop, they are responsible for calling the postprocessors if they should be called on every nonlinear iteration. In the end, the postprocessors are still only called once for the IMPES scheme (only here, instead of the main loop).

bangerth commented on source/simulator/solver_schemes.cc:
```diff
@@ -0,0 +1,744 @@
+
+        /**
+         * copied from solver.cc
+         */
```
is this comment still accurate?

gassmoeller replied:
I do not know, I just copied that. I think @MFraters introduced it. It looks like a (partly modified) copy of `solve_stokes` though.

MFraters replied:
I think it is still quite accurate, but not sure whether it is necessary to state it like this here. Just for a background: it had to be copied (and slightly modified) because of the denormalisation.

bangerth commented:
Since the tests all pass, you clearly didn't break anything in this refactoring. Please merge yourself if you think you've addressed all of the comments that needed addressing.

gassmoeller commented:
I think I addressed all comments, except for the ones I answered above.
I think there are some things that can be optimized in the newton solver scheme, but let me test that in a separate PR. This is ready from my side, if the tester is happy.
